android/jni/app-android.cpp

```diff
@@ -29,7 +36,351 @@
+// may be reused when restarting the game. Globals are DANGEROUS.
+
+float dp_xscale = 1.0f;
+float dp_yscale = 1.0f;
+
+InputState input_state;
+
+static bool first_lost = true;
+static std::string library_path;
+
+struct saved_state {
+	int x;
+};
+
+// This is called "engine" because it was in the sample code.
+struct engine {
+	struct android_app* app;
+
+	bool initialized;
+
+	JNIEnv* jniEnv;
+	jobject jniActivity;
+
+	struct saved_state state;
+
+	ASensorManager* sensorManager;
+	const ASensor* accelerometerSensor;
+	ASensorEventQueue* sensorEventQueue;
+
+	// TODO: Move out of here.
+	cInterfaceBase* glInterface;
+
+	// We will always be animating.
+	int animating;
+
+	int32_t width;
+	int32_t height;
+
+	jmethodID postCommand;
+};
+
+static engine android_engine;
+
+void ProcessFrame(engine *state);
+void ProcessResize(engine *state);
+/**
+* Initialize an EGL context for the current display.
+*/
+static int engine_init_display(struct engine* engine) {
+	// initialize OpenGL ES and EGL
+	ILOG("engine_init_display!");
+
+	engine->glInterface = HostGL_CreateGLInterface();
+	engine->glInterface->SetMode(MODE_DETECT);
+	if (!engine->glInterface->Create(engine->app->window, false)) {
+		ELOG("glInterface->Create failed");
+		return -1;
+	}
+	engine->glInterface->MakeCurrent();
+	
+	//eglQuerySurface(display, surface, EGL_WIDTH, &w);
+	//eglQuerySurface(display, surface, EGL_HEIGHT, &h);
+
+	// ILOG("EGL surface dim: %dx%d", w, h);
+
+	engine->width = engine->glInterface->GetBackBufferWidth();
+	engine->height = engine->glInterface->GetBackBufferHeight();
+
+	ILOG("Reported width/height: %dx%d", engine->width, engine->height);
+
+	pixel_xres = engine->width;
+	pixel_yres = engine->height;
+
+	ProcessResize(engine);
+
+	return 0;
+}
+
+/**
+* Just the current frame in the display.
+*/
+static void engine_draw_frame(struct engine* engine) {
+	if (!engine->glInterface) {
+		// No display.
+		return;
+	}
+
+	ProcessFrame(engine);
+
+	engine->glInterface->Swap();
+}
+
+/**
+* Tear down the EGL context currently associated with the display.
+*/
+static void engine_term_display(struct engine* engine) {
+	delete engine->glInterface;
+	engine->glInterface = nullptr;
+	engine->animating = 0;
+}
+
+bool ProcessPointerMotion(const AInputEvent *input_event) {
+	int pointerCount = AMotionEvent_getPointerCount(input_event);
+	bool retval = false;
+	for (int i = 0; i < pointerCount; i++) {
+		int pid = AMotionEvent_getPointerId(input_event, i);
+		int action = AMotionEvent_getAction(input_event);
+		int actionMasked = action & AMOTION_EVENT_ACTION_MASK;
+		int actionIndex = (action & AMOTION_EVENT_ACTION_POINTER_INDEX_MASK) >> AMOTION_EVENT_ACTION_POINTER_INDEX_SHIFT;
+		int flags = 0;
+		switch (action & AMOTION_EVENT_ACTION_MASK) {
+		case AMOTION_EVENT_ACTION_DOWN:
+		case AMOTION_EVENT_ACTION_POINTER_DOWN:
+			if (actionIndex == i) {
+				flags = TOUCH_DOWN;
+			}
+			break;
+		case AMOTION_EVENT_ACTION_UP:
+		case AMOTION_EVENT_ACTION_POINTER_UP:
```

Comment:
Should we count `AMOTION_EVENT_ACTION_CANCEL` as UP also?  Technically we should not count it as a "tap".

Also `AMOTION_EVENT_ACTION_OUTSIDE` - like a `MOVE` maybe?

-[Unknown]